### DISCUSSION TITLE
Automating AUR deployments

### DISCUSSION THREAD

**Discussion by dyc3** [dyc3/steamguard-cli#141] [General]
This thread is to figure out what exactly needs to be done to automate AUR deployments for steamguard-cli. Once we figure that out, we can open issues for the tasks that need to be done. Ideally, once we are done setting this up we should never have to touch it again.

I've attempted to automate this in the past, and one of the main blockers was that Github Actions uses ubuntu, and I have not found a way to get `makepkg` to run. I've had other problems, but I've forgotten what they were because I didn't write them down. 

I'm leaning towards using a docker container to do this. It would also be great to automatically generate the PKGBUILD from the info in `cargo.toml`

**Comment by Slips-PC** [comment]
Sorry for dropping in late.

You don't need to automatically generate the pkgver. The purpose of the pkgver function is to do this on install, and most AUR helpers will handle it properly all the same.

**Reply by Slips-PC** [reply]
If you're packaging VCS software like this you should rarely ever need to actually touch the PKGBUILD.

**Reply by dyc3** [reply]
Sorry this took way too long to follow up on.

So looking at our current PKGBUILD, the `pkgver=0.4.3.r1.145c03e9` line is redundant?

**Reply by Slips-PC** [reply]
It's fine lol.
Anyways, setting a specific `pkgver` is not *redundant*, but it doesn't need to be updated on every new push to master. In fact, you should not touch the pkgver variable after setting it, if you're using VCS package management. Not only is it a waste of valuable time (even if you are running it on a CI pipeline), it can and will screw up software that checks updates like `paru`.

**Reply by dyc3** [reply]
So setting pkgver, the variable, serves a different purpose than pkgver, the function? What should it be set to? This is very confusing. Do you have any good PKGBUILD examples?

**Reply by Slips-PC** [reply]
Alright, so basically, `pkgver`, the function, is used when you want pacman to automatically calculate the new pkgver, which in this case, you *do*. Take for example this basic `pkgver` function:
```
pkgver() {
        cd "$srcdir/${_pkgname}"
        printf "r%s.%s" "$(git rev-list --count HEAD)" "$(git rev-parse --short HEAD)"
}
```
This is used for one of my AUR packages, and basically what it does is it sets the pkgver to a specific combination of strings that denote where the HEAD is.
The `pkgver` variable should only need to be manipulated once if you're packaging for git like you currently are. So, in my case, I have the `pkgver` set to `pkgver=r39.7d20037`, which i set a while ago, when the pkgver was *actually* that. So, when you run `makepkg`, pacman will parse the PKGBUILD and see that pkgver is set to this. *then*, it will see the function, which tells it "actually, pkgver is the return of this command", which, in my case, would be `r47.3b8faaf`. (At least, this is my basic understanding of how it parses PKGBUILDs.)
So, basically, you should only set the `pkgbuild` variable once, when you first push your PKGBUILD to the AUR. Then, don't touch it, and pacman should automatically handle updating to the latest HEAD itself on the user's end.

As for examples, here's a few:
https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=clementeyene - PKGBUILD for clementeyene, a non-vcs package which doesn't use the pkgver variable.
https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=pow-git - PKGBUILD for pow-git, a git package which uses the pkgver variable.
